File: app/src/lib/http.ts
/**
 *  go-webview-app-builder
 *  Create a Aurelia JavaScript application in a WebView
 *
 *  Copyright 2024, Marc S. Brooks (https://mbrooks.info)
 *  Licensed under the MIT license:
 *  http://www.opensource.org/licenses/mit-license.php
 */

// Local modules.
import {webViewBindExists} from './utils';

export type RequestBody = {
  [T: string]: string
}

export type RequestHeaders = {
  [T: string]: string
}

export type Response = {
  Status: number,
  Headers: RequestHeaders,
  Body: string
}

/**
 * Provides HTTP client methods (Go WebView bindings and SPA fallback).
 */
export class Request {

  /**
   * Make HTTP GET request to a remote source.
   *
   * @param {String} url
   *   Request URL/URI location.
   *
   * @example
   *   const response: Response = await Request.get('https://domain.com/index.html');
   *
   * @return {Promise<Response>}
   */
  async get(url: string): Promise<Response> {
    return (webViewBindExists('browser_HttpGet'))
      ? this.normalize(await window.browser_HttpGet(url))
      : this.send('GET', url);
  }

  /**
   * Make HTTP HEAD request to a remote source.
   *
   * @param {String} url
   *   Request URL/URI location.
   *
   * @example
   *   const response: Response = await Request.head('https://domain.com/index.html');
   *
   * @return {Promise<Response>}
   */
  async head(url: string): Promise<Response> {
    return (webViewBindExists('browser_HttpHead'))
      ? this.normalize(await window.browser_HttpHead(url))
      : this.send('HEAD', url);
  }

  /**
   * Make HTTP POST request to a remote source.
   *
   * @param {String} url
   *   Request URL/URI location.
   *
   * @param {Object|undefined} headers
   *   Request HTTP headers (optional).
   *
   * @param {String|Object|undefined} body
   *   Request body (optional).
   *
   * @example
   *   const response: Response = await Request.post(
   *     'https://domain.com/index.html',
   *     {
   *       'Accept': 'text/html',
   *       'Content-Type': 'application/json',
   *
   *         ..
   *     },
   *     {
   *       name1: 'value1',
   *       name2: 'value2',
   *
   *         ..
   *     }
   *   );
   *
   * @return {Promise<Response>}
   */
  async post(url: string, headers: RequestHeaders = {}, body?: RequestBody | string): Promise<Response> {
    if (webViewBindExists('browser_HttpPost')) {

      // Convert data to Go bindings supported string types (e.g. JSON).
      const _headers = JSON.stringify(headers);
      const _body = (typeof body !== 'string') ? JSON.stringify(body) : body;

      return this.normalize(await window.browser_HttpPost(url, _headers, _body));
    } else {
      return this.send('POST', url, headers, body);
    }
  }

  /**
   * Make HTTP request to a remote source (SPA fallback only).
   *
   * @param {String} method
   *   Request method (GET|HEAD|POST).
   *
   * @param {String} url
   *   Request URL/URI location.
   *
   * @param {Object|undefined} headers
   *   Request HTTP headers (optional).
   *
   * @param {String|Object|undefined} body
   *   Request body (optional).
   *
   * @return {Promise<Response>}
   */
  send(method: string, url: string, headers: RequestHeaders = {}, body?: RequestBody | string): Promise<Response> {
    let data: FormData | string;

    if (method === 'POST' && typeof body === 'object') {
      const formData = new FormData();

      for (let key in body) {
        formData.append(key, body[key]);
      }

      data = formData;

      headers = Object.assign({
        'Content-Type': 'multipart/form-data'
      }, headers);
    }

    return fetch(url, {
        method,
        body: data,
        credentials: 'include',
        headers
      })

      // Process response.
      .then(async response => {

        // .. conform to Go WebView binding format.
        return {
          Status: response.status,
          Headers: Object.fromEntries(response.headers),
          Body: await response.text()
        };
      });
  }

  /**
   * Ensure Go WebView bindings and development response match.
   *
   * @param {String} json
   *   JSON string (Request encoded).
   *
   * @return {Response}
   */
  private normalize(json): Response {
    return JSON.parse(json);
  }
}

File: lib/http.go
//
//  go-webview-app-builder
//  Create a JavaScript single-page application (SPA) in a WebView
//
//  Copyright 2024, Marc S. Brooks (https://mbrooks.info)
//  Licensed under the MIT license:
//  http://www.opensource.org/licenses/mit-license.php
//

package lib

import (
	"bytes"
	"io"
	"encoding/json"
	"log"
	"net/http"
	"net/url"
)

//
// Request declared data types.
//
type Response struct {
	Status  int
	Headers http.Header
	Body    string
}

type Request struct {
	data Response
}

//
// NewRequest creates a HTTP client instance.
//
func NewRequest() *Request {
	request := &Request{}
	return request
}

//
// Make HTTP GET request to a remote source.
//
func (request *Request) Get(URL string) *Request {
	validateURL(URL)

	resp, err := http.Get(URL)

	if err != nil {
		log.Fatal(err)
	}

	defer resp.Body.Close()

	body, err := io.ReadAll(io.Reader(resp.Body))

	if err != nil {
		log.Fatal(err)
	}

	request.data = Response {
		Status:  resp.StatusCode,
		Headers: resp.Header,
		Body:    string(body),
	}

	return request
}

//
// Make HTTP Head request to a remote source.
//
func (request *Request) Head(URL string) *Request {
	validateURL(URL)

	resp, err := http.Head(URL)

	if err != nil {
		log.Fatal(err)
	}

	defer resp.Body.Close()

	request.data = Response {
		Status:  resp.StatusCode,
		Headers: resp.Header,
		Body:    "",
	}

	return request
}

//
// Make HTTP POST request to a remote source.
//
func (request *Request) Post(URL string, headers string, body string) *Request {
	validateURL(URL)

	buffer := bytes.NewBuffer([]byte(body))

	req, err := http.NewRequest("POST", URL, buffer)

	if err != nil {
		log.Fatal(err)
	}

	// Add HTTP headers from JSON
	for k, v := range parseJSON(headers) {
		req.Header.Set(k, v.(string))
	}

	resp, err := http.DefaultClient.Do(req)

	if err != nil {
		log.Fatal(err)
	}

	defer resp.Body.Close()

	_body, err := io.ReadAll(io.Reader(resp.Body))

	if err != nil {
		log.Fatal(err)
	}

	request.data = Response {
		Status:  resp.StatusCode,
		Headers: resp.Header,
		Body:    string(_body),
	}

	return request
}

//
// Returns unmashaled Request response.
//
func (request *Request) Response() Response {
	return request.data
}

//
// Returns Request response as JSON.
//
func (request *Request) JsonResponse() string {
	b, err := json.Marshal(request.data)

	if err != nil {
		log.Fatal(err)
	}

	return string(b)
}

//
// Converts a dynamic JSON string to a map.
//
func parseJSON(rawJSON string) map[string]interface{} {
	var data interface{}

	err := json.Unmarshal([]byte(rawJSON), &data)

	if err != nil {
		log.Fatal(err)
	}

	return data.(map[string]interface{})
}

//
// Checks for a valid URL structure, fail otherwise.
//
func validateURL(rawURL string) {
	_, err := url.ParseRequestURI(rawURL)

	if err != nil {
		log.Fatal("Invalid URL: ", err)
	}
}
